src/app/clusters/bindings/BindingManager.h

```diff
@@ -0,0 +1,180 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/CASESessionManager.h>
+#include <app/server/Server.h>
+#include <app/util/binding-table.h>
+
+namespace chip {
+
+/**
+ * Application callback function when a cluster associated with a bound peer changes.
+ *
+ * The connection is managed by the stack and peer_device is guaranteed to be available.
+ * The application shall decide the content to be sent to the peer.
+ *
+ * E.g. The application will send on/off commands to peer for the OnOff cluster.
+ *
+ */
+using BoundDeviceChangedHandler = void (*)(EndpointId localEndpoint, EndpointId remoteEndpoint, ClusterId clusterId,
+                                           DeviceProxy * peer_device, Optional<GroupId> group, void * context);
+
+/**
+ *
+ * The BindingManager class manages the connections for unicast bindings and notifies the application
+ * when a binding is ready to be communicated with.
+ *
+ * A CASE connection will be triggered when:
+ *  - The binding cluster adds a unicast entry to the binding table.
+ *  - A watched cluster changes with a unicast binding but we cannot find an active connection to the peer.
+ *
+ * The class uses an LRU mechanism to choose the connection to eliminate when there is no space for a new connection.
+ * The BindingManager class will not actively re-establish connection and will connect on-demand (when binding cluster
+ * or watched cluster is changed).
+ *
+ */
+class BindingManager
+{
+public:
+    BindingManager() :
+        mOnConnectedCallback(HandleDeviceConnected, this), mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+    {}
+
+    void RegisterBoundDeviceChangedHandler(BoundDeviceChangedHandler handler) { mBoundDeviceChangedHandler = handler; }
+
+    void SetAppServer(Server * appServer) { mAppServer = appServer; }
+
+    /*
+     * Adds a unicast binding entry.
+     *
+     */
+    CHIP_ERROR CreateUnicastBinding(FabricIndex fabric, NodeId node);
+
+    /*
+     * Totally disconnect a device. Call this function only when no binding table entry is associated with the node.
+     *
+     */
+    CHIP_ERROR RemoveUnicastBinding(FabricIndex fabric, NodeId node);
```

Comment:
```suggestion
     * Notify the BindingManager that the last unicast binding for the given node has been removed.
     *
     */
    CHIP_ERROR LastUnicastBindingRemoved(FabricIndex fabric, NodeId node);
```
What this does with connections is an implementation detail.